src/main/java/org/sqlite/core/NativeDB.java

```diff
@@ -559,4 +559,24 @@ long getUpdateListener() {
     long getProgressHandler() {
         return progressHandler;
     }
+
+    @Override
+    public synchronized native long serializeSize(String schema) throws SQLException;
+
+    @Override
+    public synchronized byte[] serialize(String schema) throws SQLException {
+        long size = serializeSize(schema);
+        if (size > Integer.MAX_VALUE) {
+            // Extra precaution as the sqlite side is also limited to 2gb.
+            throw new IllegalStateException("Database is bigger than byte[] max size: " + size);
+        }
+        byte[] buff = new byte[(int) size];
```

Comment:
It seems to me that it would be better to allocate this array in JNI and return it, and call serialize once.
Even though this DB can't be manipulated here due to the synchronize, there's nothing stopping another connection (or even another process) from manipulating the DB between these calls.

I don't think this is critical, as the JNI code in serialize double checks this and as you said, this is really more for inmemory databases. But it seems minor to fix and would make this more robust.

Reply:
> I found no way to this other than allocating DirectByteBuffer in jni
Why not use `NewByteArray()`? It should yield the same performance as allocating the byte array in java. See: `https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/functions.html#New_PrimitiveType_Array_routines`
You can change the native function to return `byte[]`.

For InMemory DBs, concurrent modification obviously isn't an issue, but this may eventually be used for file dbs.

Reply:
Done as you suggested, wasn't aware. Benchmarked, all is well. I also removed the serializedSize method. Don't need it + worried that someone will use it to find out the db size which is very wasteful compared to querying sqlite.

Reply:
Looks good to me. Nice work (: